hwy/ops/generic_ops-inl.h

```diff
@@ -91,6 +91,86 @@ HWY_API Vec<D> Inf(D d) {
   return BitCast(d, Set(du, max_x2 >> 1));
 }
 
+// ------------------------------ ZeroExtendResizeBitCast
+
+// The implementation of detail::ZeroExtendResizeBitCast for the HWY_EMU128
```

Comment:
Did we consider using the same 'include guard' technique like HWY_NATIVE_LOAD_STORE_INTERLEAVED here so that we aren't hard-coding EMU128/SCALAR? Or do you think it very unlikely that any other target would also want to implement this op themselves?

Reply:
> Did we consider using the same 'include guard' technique like HWY_NATIVE_LOAD_STORE_INTERLEAVED here so that we aren't hard-coding EMU128/SCALAR? Or do you think it very unlikely that any other target would also want to implement this op themselves?

It is very likely that targets other than HWY_EMU128/HWY_SCALAR will use the implementations of detail::ZeroExtendResizeBitCast in hwy/ops/generic_ops-inl.h for at least some of the ZeroExtendResizeBitCast cases.

The detail::ZeroExtendResizeBitCast function has hwy::SizeTag parameters for the from vector length (in bytes) and the to vector length (in bytes) to allow the detail::ZeroExtendResizeBitCast to be specialized for a few specific cases on the HWY_AVX2/HWY_AVX3/HWY_WASM_EMU256 targets.

While it is possible to use the detail::ZeroExtendResizeBitCast implementation in hwy/ops/generic_ops-inl.h on the HWY_EMU128 target, the implementation of detail::ZeroExtendResizeBitCast in hwy/ops/emu128-inl.h is used for all of the ZeroExtendResizeBitCast cases on the HWY_EMU128 target to avoid the unnecessary zero-out of the upper bytes as the ResizeBitCast implementation in hwy/ops/emu128_inl.h already does the required zero-out of the upper bytes of the result vector.

The detail::ZeroExtendResizeBitCast implementation in hwy/ops/generic_ops-inl.h cannot be used on the HWY_SCALAR target as all of the vector types on the HWY_SCALAR target have only 1 lane. The detail::ZeroExtendResizeBitCast implementation on the HWY_SCALAR target is in hwy/ops/scalar-inl.h, and the implementation of detail::ZeroExtendResizeBitCast in hwy/ops/scalar-inl.h can handle the case of a ZeroExtendResizeBitCast from a smaller lane type to a bigger lane type.

Here is the implementation of ZeroExtendResizeBitCast in hwy/ops/generic_ops-inl.h, shared across all targets, which delegates to detail::ZeroExtendResizeBitCast:
https://github.com/google/highway/blob/04aea95474744df7cb09fc209886b590136846d7/hwy/ops/generic_ops-inl.h#L159-L172

Here are the specialized cases of detail::ZeroExtendResizeBitCast that are implemented for the HWY_AVX2/HWY_AVX3/HWY_WASM_EMU256 targets (with the remaining cases of detail::ZeroExtendResizeBitCast on the HWY_AVX2/HWY_AVX3/HWY_WASM_EMU256 targets using the implementations of detail::ZeroExtendResizeBitCast in hwy/ops/generic_ops-inl.h):
https://github.com/google/highway/blob/04aea95474744df7cb09fc209886b590136846d7/hwy/ops/x86_256-inl.h#L2938-L2949
https://github.com/google/highway/blob/04aea95474744df7cb09fc209886b590136846d7/hwy/ops/x86_512-inl.h#L2451-L2501
https://github.com/google/highway/blob/04aea95474744df7cb09fc209886b590136846d7/hwy/ops/wasm_256-inl.h#L1147-L1159

Note that the detail::ZeroExtendResizeBitCast implementations in hwy/ops/x86_256-inl.h, hwy/ops/x86_512-inl.h, and hwy/ops/wasm_256-inl.h have at least one hwy::SizeTag parameter with a specific size so that the implementations of detail::ZeroExtendResizeBitCast in hwy/ops/x86_256-inl.h, hwy/ops/x86_512-inl.h, and hwy/ops/wasm_256-inl.h will be treated as more specialized than the detail::ZeroExtendResizeBitCast implementations in hwy/ops/generic_ops-inl.h.

It is also unlikely that targets other than HWY_EMU128/HWY_SCALAR would completely replace the implementations of detail::ZeroExtendResizeBitCast in hwy/ops/generic_ops-inl.h with target-specific implementations of detail::ZeroExtendResizeBitCast that cover all of the possible ZeroExtendResizeBitCast cases.

If a new target other than HWY_EMU128/HWY_SCALAR/HWY_AVX2/HWY_AVX3/HWY_WASM_EMU256 needs to specialize the detail::ZeroExtendResizeBitCast function, such new targets would likely specialize only certain detail::ZeroExtendResizeBitCast cases with target-specific implementations that have at least one hwy::SizeTag parameter with a specific size and delegate to the default detail::ZeroExtendResizeBitCast implementations in hwy/ops/generic_ops-inl.h for the remaining cases.

Reply:
Thanks for explaining. I understand that those two targets are special and thus we can hard-code them in the if, and that we already have an extension point for other targets to use.